Adopt CAPI v2beta2 status changes and respective controller changes for IBMPowerVSCluster

Karthik-K-N opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:

This PR consists of following changes

1. Adopt CAPI v1beta2 status changes
2. Standardize the controller flow by utilising CAPI provided helper functions
3. Removed usage of deprecated github.com/pkg/errors package.


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2264

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Adopt CAPI v1beta2 status changes into IBMPowerVSCluster object.
```

Karthik-K-N commented:
IBMPowerVSCluster object's status field looks like this on various stages now.

```
  v1beta2:
    Conditions:
      Last Transition Time:  2025-04-16T06:35:36Z
      Message:               
        * NetworkReady: Condition not yet reported
        * LoadBalancerReady: Condition not yet reported
        * TransitGatewayReady: Condition not yet reported
      Observed Generation:   1
      Reason:                ReadyUnknown
      Status:                Unknown
      Type:                  Ready
      Last Transition Time:  2025-04-16T06:35:39Z
      Message:
      Observed Generation:   1
      Reason:                Ready
      Status:                True
      Type:                  VPCReady
      Last Transition Time:  2025-04-16T06:35:45Z
      Message:
      Observed Generation:   1
      Reason:                Ready
      Status:                True
      Type:                  VPCSecurityGroupReady
      Last Transition Time:  2025-04-16T06:35:45Z
      Message:
      Observed Generation:   1
      Reason:                Ready
      Status:                True
      Type:                  VPCSubnetReady
      Last Transition Time:  2025-04-16T06:38:49Z
      Message:
      Observed Generation:   1
      Reason:                Ready
      Status:                True
      Type:                  WorkspaceReady
      Last Transition Time:  2025-04-16T06:35:23Z
      Message:
      Observed Generation:   1
      Reason:                NotPaused
      Status:                False
      Type:                  Paused


  v1beta2:
    Conditions:
      Last Transition Time:  2025-04-16T06:46:20Z
      Message:
      Observed Generation:   2
      Reason:                Ready
      Status:                True
      Type:                  Ready
      Last Transition Time:  2025-04-16T06:45:56Z
      Message:
      Observed Generation:   2
      Reason:                Ready
      Status:                True
      Type:                  LoadBalancerReady
      Last Transition Time:  2025-04-16T06:46:17Z
      Message:
      Observed Generation:   2
      Reason:                Ready
      Status:                True
      Type:                  NetworkReady
      Last Transition Time:  2025-04-16T06:42:52Z
      Message:
      Observed Generation:   2
      Reason:                Ready
      Status:                True
      Type:                  TransitGatewayReady
      Last Transition Time:  2025-04-16T06:35:39Z
      Message:
      Observed Generation:   2
      Reason:                Ready
      Status:                True
      Type:                  VPCReady
      Last Transition Time:  2025-04-16T06:35:45Z
      Message:
      Observed Generation:   2
      Reason:                Ready
      Status:                True
      Type:                  VPCSecurityGroupReady
      Last Transition Time:  2025-04-16T06:35:45Z
      Message:
      Observed Generation:   2
      Reason:                Ready
      Status:                True
      Type:                  VPCSubnetReady
      Last Transition Time:  2025-04-16T06:38:49Z
      Message:
      Observed Generation:   2
      Reason:                Ready
      Status:                True
      Type:                  WorkspaceReady
      Last Transition Time:  2025-04-16T06:35:23Z
      Message:
      Observed Generation:   2
      Reason:                NotPaused
      Status:                False
      Type:                  Paused





v1beta2:
    Conditions:
      Last Transition Time:  2025-04-16T08:29:23Z
      Message:               * LoadBalancerReady: Deleting
* TransitGatewayReady: Deleting
      Observed Generation:   3
      Reason:                NotReady
      Status:                False
      Type:                  Ready
      Last Transition Time:  2025-04-16T08:32:39Z
      Message:
      Observed Generation:   3
      Reason:                Deleting
      Status:                False
      Type:                  LoadBalancerReady
```

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1183,24 +1195,26 @@ func (s *PowerVSClusterScope) createVPC() (*string, error) {
 	// check whether user has set the vpc subnets
 	if len(s.IBMPowerVSCluster.Spec.VPCSubnets) == 0 {
 		// if the user did not set any subnet, we try to create subnet in all the zones.
+		log.V(3).Info("VPC subnets are not set, constructing one")
```
```suggestion
		log.V(3).Info("VPC subnets details are not set in spec, creating subnets in all zones in the given region")
```

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1714,55 +1732,57 @@ func (s *PowerVSClusterScope) validateVPCSecurityGroup(securityGroup infrav1beta
+		requeue, err := s.checkAndUpdateTransitGateway(ctx, tg)
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("failed to check and update transit gateway exists: %w", err)
```
```suggestion
			return false, fmt.Errorf("failed to check and update connections if transit gateway exists: %w", err)
```

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1864,16 +1886,16 @@ func (s *PowerVSClusterScope) checkAndUpdateTransitGatewayConnections(transitGat
-			return false, err
+		log.V(3).Info("Only VPC connection not exist in transit gateway, creating it")
+		if err := s.createTransitGatewayConnection(ctx, transitGateway.ID, ptr.To(getTGVPCConnectionName(*transitGateway.Name)), vpcCRN, vpcNetworkConnectionType); err != nil {
+			return false, fmt.Errorf("failed to create VPC transit gateway connections: %w", err)
```
```suggestion
			return false, fmt.Errorf("failed to create VPC transit gateway connection: %w", err)
```

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1864,16 +1886,16 @@ func (s *PowerVSClusterScope) checkAndUpdateTransitGatewayConnections(transitGat
-			return false, err
+		log.V(3).Info("Only PowerVS connection not exist in transit gateway, creating it")
+		if err := s.createTransitGatewayConnection(ctx, transitGateway.ID, ptr.To(getTGPowerVSConnectionName(*transitGateway.Name)), pvsServiceInstanceCRN, powervsNetworkConnectionType); err != nil {
+			return false, fmt.Errorf("failed to create PowerVS transit gateway connections: %w", err)
```
```suggestion
			return false, fmt.Errorf("failed to create PowerVS transit gateway connection: %w", err)
```

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -2064,35 +2087,35 @@ func (s *PowerVSClusterScope) ReconcileLoadBalancers() (bool, error) {
+		loadBalancerStatus, err := s.checkLoadBalancer(ctx, loadBalancer)
 		if err != nil {
-			return false, err
+			return false, fmt.Errorf("failed to check load balancer: %w", err)
```
```suggestion
			return false, fmt.Errorf("failed to check if load balancer exists: %w", err)
```

Amulyam24 commented on cloud/scope/powervs_cluster.go:
```diff
@@ -2294,16 +2320,15 @@ func (s *PowerVSClusterScope) ReconcileCOSInstance() error {
 	} else if err != nil {
-		s.Error(err, "failed to check COS bucket")
-		return err
+		return fmt.Errorf("failed to check COS bucket: %w", err)
```
```suggestion
		return fmt.Errorf("failed to check if COS bucket exists: %w", err)
```

Amulyam24 commented on controllers/ibmpowervsmachine_controller.go:
```diff
@@ -144,8 +144,9 @@ func (r *IBMPowerVSMachineReconciler) Reconcile(ctx context.Context, req ctrl.Re
 
 	// Fetch the IBMPowerVSImage.
-	ibmPowerVSImage := &infrav1beta2.IBMPowerVSImage{}
+	var ibmPowerVSImage *infrav1beta2.IBMPowerVSImage
```
This change has been merged, can you rebase the PR?

Amulyam24 commented:
Not to be handled in this PR but for future, we should consider
1. Move naming to PowerVS workspace instead of service instance to keep up with the cloud terms
2. Error format is incosistent with `error fetching/creating` and `failed to fetch/create`, should follow one format
wdyt?

Karthik-K-N commented:
> Not to be handled in this PR but for future, we should consider
> 
>     1. Move naming to PowerVS workspace instead of service instance to keep up with the cloud terms

Yes, We are planning to do the same in our next api version where we plan to support multi-workspace,

> 
>     2. Error format is incosistent with `error fetching/creating` and `failed to fetch/create`, should follow one format
>        wdyt?


Yes, Agree, next iteration will fix the correct verb

Prajyot-Parab commented on controllers/ibmpowervscluster_controller.go:
```diff
@@ -407,105 +521,147 @@ func (r *IBMPowerVSClusterReconciler) reconcileDelete(ctx context.Context, clust
+		allErrs = append(allErrs, fmt.Errorf("failed to delete transit gateway: %w", err))
 	} else if requeue {
-		clusterScope.Info("Cleaning up transit gateway is pending, requeuing")
+		log.Info("Deleting transit gateway is pending, requeuing")
```
```suggestion
		log.Info("Transit gateway deletion is pending, requeuing")
```

Prajyot-Parab reviewed (changes requested):
Minor nits.
overall lgtm.

Prajyot-Parab commented:
/lgtm
/hold for @Amulyam24

dharaneeshvrd reviewed (commented):
Minor nits on logging
Few places you have reduced the logging level, please recheck them once
When ever we check resource in cloud, we say found in cloud. IMO just found also makes sense. e.g.: Found PowerVS service instance in cloud -> Found PowerVS service instance

Karthik-K-N commented on cloud/scope/powervs_cluster.go:
```diff
@@ -1297,7 +1310,7 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
 	vpcID := s.GetVPCID()
 	if vpcID == nil {
-		return nil, fmt.Errorf("VPC ID is empty")
+		return nil, fmt.Errorf("VPC id is empty")
```
will make it uniform and keep it in small "id"

Amulyam24 replied:
Smallcase `id` for variable names makes sense but I feel it is better to keep it as `ID` in logs. wdyt?

Karthik-K-N replied:
Yeah, In Logs we can keep ID

Karthik-K-N commented:
> Minor nits on logging Few places you have reduced the logging level, please recheck them once When ever we check resource in cloud, we say found in cloud. IMO just found also makes sense. e.g.: Found PowerVS service instance in cloud -> Found PowerVS service instance

Thank you for the revew, I have addressed all your comments.

Since we check for id of various resources  in status, previously we kept "found in cloud" just to indicate that we are looking specifically in cloud.

Amulyam24 commented:
/retest

Amulyam24 reviewed (commented):
minor nit, otherwise LGTM!

Prajyot-Parab commented on api/v1beta2/conditions_consts.go:
```diff
@@ -180,3 +180,118 @@ const (
+	IBMPowerVSClusterNotReadyV1Beta2Reason = capiv1beta1.NotReadyV1Beta2Reason
+
+	// IBMPowerVSClusterReadyUnknownV1Beta2Reason surfaces when at least one of the IBMPowerVSCluster readiness criteria is unknown
+	// and none of the IBMPowerVSCluster readiness criteria is not met.
```
```suggestion
	// and none of the IBMPowerVSCluster readiness criteria is met.
```

Prajyot-Parab commented:
/unhold
